sky/shell/tracing_controller.cc

```diff
@@ -155,13 +182,97 @@ void TracingController::StartTracing() {
   StartBaseTracing();
 }
 
-void TracingController::StopTracing(const base::FilePath& path) {
+void TracingController::StopTracing(const base::FilePath& path,
+                                    const base::Closure& on_done) {
   LOG(INFO) << "Saving trace to " << path.LossyDisplayName();
 
-  trace_file_ = std::unique_ptr<base::File>(new base::File(
-      path, base::File::FLAG_CREATE_ALWAYS | base::File::FLAG_WRITE));
+  auto file = std::make_shared<base::File>(
+      path, base::File::FLAG_CREATE_ALWAYS | base::File::FLAG_WRITE);
+
+  // Make world readable so root access is not required on Android to view the
+  // trace.
   base::SetPosixFilePermissions(path, base::FILE_PERMISSION_MASK);
-  StopBaseTracing();
+
+  StopTracing(file, on_done);
+}
+
+void TracingController::StopTracing(std::shared_ptr<base::File> file,
+                                    const base::Closure& on_done) {
+  StopBaseTracing(file, on_done);
+}
+
+void TracingController::CaptureObservatoryTraceInWorker(
+    const base::Closure& on_done) {
+  // This is on a worker pool thread.
+  observatory_trace_ = std::make_shared<base::File>(
+      ObservatoryExtraTraceDataPath(),
+      base::File::FLAG_CREATE_ALWAYS | base::File::FLAG_WRITE);
+  StopTracing(observatory_trace_, on_done);
+}
+
+void TracingController::StartTracingFromObservatory() {
+  StartBaseTracing();
+}
+
+void TracingController::StopTracingFromObservatory() {
+  base::trace_event::TraceLog::GetInstance()->SetDisabled();
+
+  std::lock_guard<std::mutex> lock(observatory_trace_mutex_);
+  base::DeleteFile(ObservatoryExtraTraceDataPath(), false);
+  observatory_trace_ = nullptr;
+}
+
+static std::string ReadTraceFileAsString(base::File* file) {
+  if (file == nullptr) {
+    return nullptr;
+  }
+
+  std::stringstream stream;
+  const size_t read_block = 4096;
+  char* buffer = static_cast<char*>(calloc(read_block, sizeof(char)));
+  int64 offset = 0;
+
+  while (true) {
+    int read = file->Read(offset, buffer, read_block);
+
+    if (read <= 0) {
+      break;
+    }
+
+    offset += read;
+    stream.write(buffer, read);
+  }
+
+  free(buffer);
+
+  return stream.str();
+}
+
+static void LatchCountDown(std::shared_ptr<CountdownLatch> latch) {
```

Comment:
Please don't use `shard_ptr`.  It's not a well-designed class.  Consider scoped_refptr instead.